pkg/storage/redis/redis.go

```diff
@@ -0,0 +1,241 @@
+				}
+				return c, nil
+			},
+			TestOnBorrow: func(c redis.Conn, t time.Time) error {
+				if time.Since(t) < time.Minute {
+					return nil
+				}
+				_, err := c.Do("PING")
+				if err != nil {
+					return fmt.Errorf(`c.Do("PING"): %w`, err)
+				}
+				return nil
+			},
+		},
+		deletePermanentlyAfter: int64(internal_databroker.DefaultDeletePermanentlyAfter.Seconds()),
+		recordType:             recordType,
+		versionSet:             "version_set",
+		deletedSet:             "deleted_set",
+	}
+	return db, nil
+}
+
+// Put sets new record for given id with input data.
+func (db *DB) Put(ctx context.Context, id string, data *anypb.Any) error {
+	c := db.pool.Get()
+	defer c.Close()
+	record := db.Get(ctx, id)
+	if record == nil {
+		record = new(databroker.Record)
+		record.CreatedAt = ptypes.TimestampNow()
+	}
+
+	record.Data = data
+	record.ModifiedAt = ptypes.TimestampNow()
+	record.Type = db.recordType
+	record.Id = id
+	record.Version = fmt.Sprintf("%012X", atomic.AddUint64(&db.lastVersion, 1))
+	b, err := proto.Marshal(record)
+	if err != nil {
+		return err
+	}
+	cmds := map[string][]interface{}{
+		"MULTI": nil,
+		"HSET":  {db.recordType, id, string(b)},
+		"ZADD":  {db.versionSet, db.lastVersion, id},
+	}
+	if _, err := db.tx(c, cmds); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Get retrieves a record from redis.
+func (db *DB) Get(_ context.Context, id string) *databroker.Record {
+	c := db.pool.Get()
+	defer c.Close()
+
+	b, err := redis.Bytes(c.Do("HGET", db.recordType, id))
+	if err != nil {
+		return nil
+	}
+
+	return db.toPbRecord(b)
+}
+
+// GetAll retrieves all records from redis.
+func (db *DB) GetAll(ctx context.Context) []*databroker.Record {
+	return db.getAll(ctx, func(record *databroker.Record) bool { return true })
+}
+
+// List retrieves all records since given version.
+//
+// "version" is in hex format, invalid version will be treated as 0.
+func (db *DB) List(ctx context.Context, sinceVersion string) []*databroker.Record {
+	c := db.pool.Get()
+	defer c.Close()
+
+	v, err := strconv.ParseUint(sinceVersion, 16, 64)
+	if err != nil {
+		v = 0
+	}
+
+	ids, err := redis.Strings(c.Do("ZRANGEBYSCORE", db.versionSet, fmt.Sprintf("(%d", v), "+inf"))
+	if err != nil {
+		return nil
+	}
+
+	records := make([]*databroker.Record, 0, len(ids))
+	for _, id := range ids {
+		b, err := redis.Bytes(c.Do("HGET", db.recordType, id))
+		if err != nil {
+			continue
+		}
+		records = append(records, db.toPbRecord(b))
+	}
+	return records
+}
+
+// Delete sets a record DeletedAt field and set its TTL.
+func (db *DB) Delete(ctx context.Context, id string) error {
+	c := db.pool.Get()
+	defer c.Close()
+
+	r := db.Get(ctx, id)
+	if r == nil {
+		return errors.New("not found")
+	}
+	r.DeletedAt = ptypes.TimestampNow()
+	b, err := proto.Marshal(r)
+	if err != nil {
+		return err
+	}
+	cmds := map[string][]interface{}{
+		"MULTI": nil,
+		"HSET":  {db.recordType, id, string(b)},
+		"SADD":  {db.deletedSet, id},
```

Comment:
I think we need to do a ZADD and increment the version?

Reply:
@calebdoxsey why 🤔 What is doing here is equivalent with what the `inmemory` backend does.

Reply:
I think if you delete a record, List won't pick up the change.

Reply:
`List` will pick up if the key is still not deleted by `ClearDeleted` right? So I think it should be the right behavior.

Reply:
@calebdoxsey isn't that we're checking that `DeletedAt` is not nil to detect this case? At least in `inmemory` implementation., I don't see it increase the version when delete one.

If that's the bug, I will fix in another PR.

Reply:
The in memory version changes the version number when you call `Delete`.

```
record.Version = fmt.Sprintf("%012X", atomic.AddUint64(&db.lastVersion, 1))
```

Reply:
@calebdoxsey ah ok, missing that part, Delete call replaceOrInsert as well

Reply:
@calebdoxsey fixed.